modules/@apostrophecms/login/index.js

```diff
@@ -106,34 +107,118 @@ module.exports = {
     return {
       post: {
         async login(req) {
-          const username = self.apos.launder.string(req.body.username);
-          const password = self.apos.launder.string(req.body.password);
+          const earlyRequirements = Object.fromEntries(Object.entries(self.requirements).filter(([ name, requirement ]) => requirement.phase !== 'afterPasswordVerified'));
+          const lateRequirements = Object.fromEntries(Object.entries(self.requirements).filter(([ name, requirement ]) => requirement.phase === 'afterPasswordVerified'));
           const session = self.apos.launder.boolean(req.body.session);
-          if (!(username && password)) {
-            throw self.apos.error('invalid', req.t('apostrophe:loginPageBothRequired'));
-          }
-          const user = await self.apos.login.verifyLogin(username, password);
-          if (!user) {
-            // For security reasons we may not tell the user which case applies
-            throw self.apos.error('invalid', req.t('apostrophe:loginPageBadCredentials'));
-          }
-          if (session) {
-            const passportLogin = (user) => {
-              return require('util').promisify(function(user, callback) {
-                return req.login(user, callback);
-              })(user);
-            };
-            await passportLogin(user);
-          } else {
-            const token = cuid();
-            await self.bearerTokens.insert({
-              _id: token,
-              userId: user._id,
-              expires: new Date(new Date().getTime() + (self.options.bearerTokens.lifetime || (86400 * 7 * 2)) * 1000)
+          const passportLogin = (user) => {
+            return require('util').promisify(function(user, callback) {
+              return req.login(user, callback);
+            })(user);
+          };
+          // Don't make verify functions worry about whether this object
+          // is present, just the value of their own sub-property
+          req.body.requirements = req.body.requirements || {};
+
+          if (req.body.incompleteToken) {
```

Comment:
The conditional flow would be easier to follow with the logic broken into separate functions.

Reply:
Agreed, I will refactor this route. See slack for a response about the tech design and tickets etc. This ticket is purely the server side.